File: packages/barcode-scanning/android/src/main/kotlin/com/margelo/nitro/nitromlkit/barcode/HybridBarcodeScanner.kt
package com.margelo.nitro.nitromlkit.barcode

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import com.google.mlkit.vision.barcode.BarcodeScanning
import com.google.mlkit.vision.common.InputImage
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.tasks.await
import java.net.URL
import com.google.mlkit.vision.barcode.common.Barcode as MlkitBarcode

/**
 * Native Android implementation of BarcodeScanner.
 * Extends the Nitrogen-generated HybridBarcodeScannerSpec abstract class.
 *
 * Uses MLKit's bundled barcode-scanning model (all 1D/2D symbologies).
 */
class HybridBarcodeScanner : HybridBarcodeScannerSpec() {

  // Default client detects all supported formats; reused across calls.
  private val scanner by lazy { BarcodeScanning.getClient() }

  // MARK: - Spec methods

  override fun scan(imageUri: String): Promise<Array<Barcode>> {
    return Promise.async { scanImage(imageUri) }
  }

  override fun scanFirst(imageUri: String): Promise<Barcode?> {
    return Promise.async { scanImage(imageUri).firstOrNull() }
  }

  override fun scanBatch(imageUris: Array<String>, concurrency: Double): Promise<Array<BatchScanResult>> {
    return Promise.async {
      val maxConcurrent = concurrency.toInt().coerceAtLeast(1)
      val results = mutableListOf<BatchScanResult>()
      var start = 0
      while (start < imageUris.size) {
        val end = minOf(start + maxConcurrent, imageUris.size)
        val chunk = coroutineScope {
          (start until end).map { idx ->
            async {
              try {
                BatchScanResult(idx.toDouble(), scanImage(imageUris[idx]), true, null)
              } catch (e: Exception) {
                BatchScanResult(idx.toDouble(), emptyArray(), false, e.message ?: "unknown error")
              }
            }
          }.awaitAll()
        }
        results.addAll(chunk)
        start = end
      }
      results.toTypedArray()
    }
  }

  override fun isAvailable(): Boolean = true

  // MARK: - Private helpers

  private suspend fun scanImage(uri: String): Array<Barcode> {
    val bitmap = loadBitmap(uri) ?: throw Error("Failed to load image: $uri")
    val image = InputImage.fromBitmap(bitmap, 0)
    val barcodes = scanner.process(image).await()
    return barcodes.map { mapBarcode(it) }.toTypedArray()
  }

  private fun mapBarcode(b: MlkitBarcode): Barcode {
    val box = b.boundingBox
    val raw = b.rawValue ?: ""
    return Barcode(
      rawValue = raw,
      displayValue = b.displayValue ?: raw,
      format = formatName(b.format),
      valueType = valueTypeName(b.valueType),
      bounds = BarcodeBounds(
        x = (box?.left ?: 0).toDouble(),
        y = (box?.top ?: 0).toDouble(),
        width = (box?.width() ?: 0).toDouble(),
        height = (box?.height() ?: 0).toDouble(),
      ),
    )
  }

  private fun formatName(format: Int): String = when (format) {
    MlkitBarcode.FORMAT_CODE_128 -> "CODE_128"
    MlkitBarcode.FORMAT_CODE_39 -> "CODE_39"
    MlkitBarcode.FORMAT_CODE_93 -> "CODE_93"
    MlkitBarcode.FORMAT_CODABAR -> "CODABAR"
    MlkitBarcode.FORMAT_DATA_MATRIX -> "DATA_MATRIX"
    MlkitBarcode.FORMAT_EAN_13 -> "EAN_13"
    MlkitBarcode.FORMAT_EAN_8 -> "EAN_8"
    MlkitBarcode.FORMAT_ITF -> "ITF"
    MlkitBarcode.FORMAT_QR_CODE -> "QR_CODE"
    MlkitBarcode.FORMAT_UPC_A -> "UPC_A"
    MlkitBarcode.FORMAT_UPC_E -> "UPC_E"
    MlkitBarcode.FORMAT_PDF417 -> "PDF417"
    MlkitBarcode.FORMAT_AZTEC -> "AZTEC"
    else -> "UNKNOWN"
  }

  private fun valueTypeName(type: Int): String = when (type) {
    MlkitBarcode.TYPE_CONTACT_INFO -> "CONTACT_INFO"
    MlkitBarcode.TYPE_EMAIL -> "EMAIL"
    MlkitBarcode.TYPE_ISBN -> "ISBN"
    MlkitBarcode.TYPE_PHONE -> "PHONE"
    MlkitBarcode.TYPE_PRODUCT -> "PRODUCT"
    MlkitBarcode.TYPE_SMS -> "SMS"
    MlkitBarcode.TYPE_TEXT -> "TEXT"
    MlkitBarcode.TYPE_URL -> "URL"
    MlkitBarcode.TYPE_WIFI -> "WIFI"
    MlkitBarcode.TYPE_GEO -> "GEO"
    MlkitBarcode.TYPE_CALENDAR_EVENT -> "CALENDAR_EVENT"
    MlkitBarcode.TYPE_DRIVER_LICENSE -> "DRIVER_LICENSE"
    else -> "UNKNOWN"
  }

  private fun loadBitmap(uri: String): Bitmap? {
    return try {
      when {
        uri.startsWith("file://") -> BitmapFactory.decodeFile(uri.removePrefix("file://"))
        uri.startsWith("/") -> BitmapFactory.decodeFile(uri)
        uri.startsWith("content://") -> {
          val context = NitroModules.applicationContext ?: return null
          context.contentResolver.openInputStream(Uri.parse(uri))?.use { BitmapFactory.decodeStream(it) }
        }
        else -> URL(uri).openStream().use { BitmapFactory.decodeStream(it) }
      }
    } catch (e: Exception) {
      null
    }
  }
}

File: packages/barcode-scanning/ios/HybridBarcodeScanner.swift
import Foundation
import UIKit
import NitroModules
import MLKitBarcodeScanning
import MLKitVision

/**
 * Native iOS implementation of BarcodeScanner.
 * Conforms to the Nitrogen-generated HybridBarcodeScannerSpec protocol.
 *
 * Uses MLKit's bundled barcode-scanning model (all 1D/2D symbologies).
 * Mirrors the Android (Kotlin) implementation's format / valueType string
 * mapping so results are identical across platforms.
 */
class HybridBarcodeScanner: HybridBarcodeScannerSpec {

    // MARK: - HybridObject boilerplate
    var memorySize: Int { MemoryLayout<HybridBarcodeScanner>.size }

    // Disambiguate the MLKit `Barcode` class from the Nitro `Barcode` struct.
    private typealias MLBarcode = MLKitBarcodeScanning.Barcode

    // Default scanner detects all supported formats; reused across calls.
    private lazy var scanner: BarcodeScanner = BarcodeScanner.barcodeScanner()

    // MARK: - Spec methods

    func scan(imageUri: String) throws -> Promise<[Barcode]> {
        return Promise.async {
            return try await self.scanImage(imageUri)
        }
    }

    func scanFirst(imageUri: String) throws -> Promise<Barcode?> {
        return Promise.async {
            return try await self.scanImage(imageUri).first
        }
    }

    func scanBatch(imageUris: [String], concurrency: Double) throws -> Promise<[BatchScanResult]> {
        return Promise.async {
            let maxConcurrent = max(1, Int(concurrency))
            var results = [BatchScanResult]()
            results.reserveCapacity(imageUris.count)

            for chunkStart in stride(from: 0, to: imageUris.count, by: maxConcurrent) {
                let chunkEnd = min(chunkStart + maxConcurrent, imageUris.count)
                let chunk = Array(imageUris[chunkStart..<chunkEnd])

                let chunkResults = await withTaskGroup(of: (Int, BatchScanResult).self) { group in
                    for (i, uri) in chunk.enumerated() {
                        let globalIdx = chunkStart + i
                        group.addTask {
                            do {
                                let barcodes = try await self.scanImage(uri)
                                return (globalIdx, BatchScanResult(index: Double(globalIdx), barcodes: barcodes, success: true, error: nil))
                            } catch {
                                return (globalIdx, BatchScanResult(index: Double(globalIdx), barcodes: [], success: false, error: error.localizedDescription))
                            }
                        }
                    }
                    var out = [(Int, BatchScanResult)]()
                    for await result in group { out.append(result) }
                    return out.sorted(by: { $0.0 < $1.0 }).map(\.1)
                }
                results.append(contentsOf: chunkResults)
            }
            return results
        }
    }

    func isAvailable() throws -> Bool {
        return true
    }

    // MARK: - Private helpers

    private func scanImage(_ uri: String) async throws -> [Barcode] {
        guard let image = loadImage(from: uri) else {
            throw RuntimeError.error(withMessage: "Failed to load image: \(uri)")
        }
        let visionImage = VisionImage(image: image)
        visionImage.orientation = image.imageOrientation
        let mlBarcodes = try await scanner.process(visionImage)
        return mlBarcodes.map { mapBarcode($0) }
    }

    private func mapBarcode(_ b: MLBarcode) -> Barcode {
        let raw = b.rawValue ?? ""
        let box = b.frame
        return Barcode(
            rawValue: raw,
            displayValue: b.displayValue ?? raw,
            format: formatName(b.format),
            valueType: valueTypeName(b.valueType),
            bounds: BarcodeBounds(
                x: Double(box.origin.x),
                y: Double(box.origin.y),
                width: Double(box.width),
                height: Double(box.height)
            )
        )
    }

    /// Maps MLKit's `BarcodeFormat` (an OptionSet on iOS) to the same uppercase
    /// strings the Kotlin implementation emits. A detected barcode carries a
    /// single format bit, so `contains(_:)` acts as an equality check here.
    private func formatName(_ format: BarcodeFormat) -> String {
        if format.contains(.code128) { return "CODE_128" }
        if format.contains(.code39) { return "CODE_39" }
        if format.contains(.code93) { return "CODE_93" }
        if format.contains(.codaBar) { return "CODABAR" }
        if format.contains(.dataMatrix) { return "DATA_MATRIX" }
        if format.contains(.EAN13) { return "EAN_13" }
        if format.contains(.EAN8) { return "EAN_8" }
        if format.contains(.ITF) { return "ITF" }
        if format.contains(.qrCode) { return "QR_CODE" }
        if format.contains(.UPCA) { return "UPC_A" }
        if format.contains(.UPCE) { return "UPC_E" }
        if format.contains(.PDF417) { return "PDF417" }
        if format.contains(.aztec) { return "AZTEC" }
        return "UNKNOWN"
    }

    private func valueTypeName(_ type: BarcodeValueType) -> String {
        switch type {
        case .contactInfo: return "CONTACT_INFO"
        case .email: return "EMAIL"
        case .ISBN: return "ISBN"
        case .phone: return "PHONE"
        case .product: return "PRODUCT"
        case .SMS: return "SMS"
        case .text: return "TEXT"
        case .URL: return "URL"
        case .wiFi: return "WIFI"
        case .geographicCoordinates: return "GEO"
        case .calendarEvent: return "CALENDAR_EVENT"
        case .driversLicense: return "DRIVER_LICENSE"
        default: return "UNKNOWN"
        }
    }

    private func loadImage(from uri: String) -> UIImage? {
        let path: String
        if uri.hasPrefix("file://") {
            path = String(uri.dropFirst(7))
        } else if uri.hasPrefix("/") {
            path = uri
        } else {
            guard let url = URL(string: uri), let data = try? Data(contentsOf: url) else { return nil }
            return UIImage(data: data)
        }
        return UIImage(contentsOfFile: path)
    }
}
